### PR TITLE
Ptx 15684:Check error when reduce size

### PR DESCRIPTION
**Pull request by soma-purestorage**
Check error when reduce size is tried

  Make sure to have done the following:
  [] Signed off your work as per the DCO.
  [] Add unit-tests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #PTX-15684

**Special notes for your reviewer**:
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Somashekar/job/Torpedo/job/torpedo-resize1/11/console

### CODE REVIEW

**Comment by lsrinivas-pure** [comment]
@soma-purestorage please discuss this scenario with me when u get time

File: tests/basic/storage_pool_test.go [hunk 1]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+			if re.MatchString(fmt.Sprintf("%v", err)) == false {
+				errMatch = fmt.Errorf("Failed to verify failure using reduced size PoolUUID [%v]", poolUUID)
+			}
+			dash.VerifyFatal(errMatch, nil, "Pool expand with invalid Poolsize completed?")
+
+			endTime = time.Now()
+
+			// Get alerts from the cluster between startTime till endTime [ PWX-28484 ]
+			log.InfoD("Getting alerts generated by Pool between startTime : [%v] and endTime : [%v]", startMinusTenHours, endTime)
+			alerts, err := Inst().V.GetAlertsUsingResourceTypeByTime(api.ResourceType_RESOURCE_TYPE_POOL, startMinusTenHours, endTime)
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6389]: just check if you need to validate the alerts ? if not please remove the alert checking part

File: tests/basic/storage_pool_test.go [hunk 2]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
 		AfterEachTest(contexts, testrailID, runID)
 	})
 })
+
+var _ = Describe("{PoolResizeReduceErrorcheck}", func() {
+	var testrailID = 79487
+	// Testrail Description : Resize with invalid pool size less than existing
+	// Testrail Corresponds : https://portworx.testrail.net/index.php?/cases/view/84470
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6322]: please check if the testrail description and corresponds needs to be updated ?

File: tests/basic/storage_pool_test.go [hunk 3]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+	var runID int
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("PoolResizeReduceErrorcheck",
+			"Resize with invalid reduced pool size",
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6327]: please change the statement

File: tests/basic/storage_pool_test.go [hunk 4]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+	It(stepLog, func() {
+		log.InfoD(stepLog)
+
+		startTime := time.Now()
+
+		// Get alerts from ten hours before current time till current time to get the sufficient alerts
+		startMinusTenHours := startTime.Add(time.Duration(-600) * time.Minute)
+		endTime := time.Now()
+		alertsBefore, err := Inst().V.GetAlertsUsingResourceTypeByTime(api.ResourceType_RESOURCE_TYPE_POOL, startMinusTenHours, endTime)
+		log.FailOnError(err, "Failed to fetch alerts between startTime [%v] and endTime [%v]", startMinusTenHours, endTime)
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6343]: please check if you need to check for alerts ? if not please delete the lines

File: tests/basic/storage_pool_test.go [hunk 5]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+		// Get the Pool UUID on which IO is running
+		poolUUID, err := GetPoolIDWithIOs()
+		log.FailOnError(err, "Failed to get pool using UUID [%v]", poolUUID)
+		log.InfoD("Pool UUID on which IO is running [%s]", poolUUID)
+
+		nodeDetail, err := GetNodeWithGivenPoolID(poolUUID)
+		log.FailOnError(err, "Failed to get Node Details from PoolUUID [%v]", poolUUID)
+
+		// Resize Pool with lower pool size than existing
+		// Do Pool Expand on the Node
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6361]: delete this line

File: tests/basic/storage_pool_test.go [hunk 6]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+		stepLog = fmt.Sprintf("Expanding pool on node [%s] and pool UUID: [%s] using auto", nodeDetail.Name, poolUUID)
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6362]: update print statement as "Resizing pool on node [%s] and pool UUID: [%s] using auto", nodeDetail.Name, poolUUID"

**Review by vprabhakar-px — CHANGES_REQUESTED** [review]

File: tests/basic/storage_pool_test.go (hunk 1 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Done

File: tests/basic/storage_pool_test.go [hunk 7]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+			errMatch = nil
+			re := regexp.MustCompile(fmt.Sprintf("service pool expand: pool: %s is already at a size..*", poolUUID))
+			if re.MatchString(fmt.Sprintf("%v", err)) == false {
+				errMatch = fmt.Errorf("Failed to verify failure using reduced size PoolUUID [%v]", poolUUID)
+			}
+			dash.VerifyFatal(errMatch, nil, "Pool expand with invalid Poolsize completed?")
+
+			endTime = time.Now()
+
+			// Get alerts from the cluster between startTime till endTime [ PWX-28484 ]
```
[the remark this answers is not in the document]
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Removed

File: tests/basic/storage_pool_test.go (hunk 4 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: removed

File: tests/basic/storage_pool_test.go (hunk 6 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go:6354]: Done

File: tests/basic/storage_pool_test.go (hunk 5 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go:6353]: Done

File: tests/basic/storage_pool_test.go (hunk 3 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Done

File: tests/basic/storage_pool_test.go (hunk 2 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: this i have to check with leela

File: tests/basic/storage_pool_test.go [hunk 8]
```diff
@@ -6315,3 +6315,88 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+			err = Inst().V.ExpandPoolUsingPxctlCmd(*nodeDetail, poolUUID, api.SdkStoragePool_RESIZE_TYPE_AUTO, expectedSize)
+
+			// Verify error on pool expansion failure
+			var errMatch error
+			errMatch = nil
+			re := regexp.MustCompile(fmt.Sprintf("service pool expand: pool: %s is already at a size..*", poolUUID))
+			if re.MatchString(fmt.Sprintf("%v", err)) == false {
+				errMatch = fmt.Errorf("Failed to verify failure using reduced size PoolUUID [%v]", poolUUID)
+			}
+			dash.VerifyFatal(errMatch, nil, "Pool expand with invalid Poolsize completed?")
```
[the remark this answers is not in the document]
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Changed

File: tests/basic/storage_pool_test.go (hunk 2 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Done

File: tests/basic/storage_pool_test.go [hunk 9]
```diff
@@ -6315,3 +6315,68 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+		contexts = make([]*scheduler.Context, 0)
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			contexts = append(contexts, ScheduleApplications(fmt.Sprintf("reducesize-%d", i))...)
+		}
+		ValidateApplications(contexts)
+		defer appsValidateAndDestroy(contexts)
+
+		// Get the Pool UUID on which IO is running
+		poolUUID, err := GetPoolIDWithIOs()
+		log.FailOnError(err, "Failed to get pool using UUID [%v]", poolUUID)
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6342]: you might not get pool UUID during err , 
please update it as log.FailOnError(err, "Failed to get pool using UUID")

**Review by vprabhakar-px — CHANGES_REQUESTED** [review]

File: tests/basic/storage_pool_test.go (hunk 9 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: DOne

File: tests/basic/storage_pool_test.go [hunk 10]
```diff
@@ -6315,3 +6315,68 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+
+		// Resize Pool with lower pool size than existing
+		// Do Pool Expand on the Node
+		stepLog = fmt.Sprintf("Resizing pool on node [%s] and pool UUID: [%s] using auto", nodeDetail.Name, poolUUID)
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			poolToBeResized, err := GetStoragePoolByUUID(poolUUID)
+			log.FailOnError(err, "Failed to get pool using UUID [%s]", poolUUID)
+
+			log.FailOnError(err, "error getting drive size for pool [%s]", poolToBeResized.Uuid)
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6356]: Line number 6354 is not needed , we are checking for poolresize error in line 5356

File: tests/basic/storage_pool_test.go [hunk 11]
```diff
@@ -6315,3 +6315,68 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+		// Get the Pool UUID on which IO is running
+		poolUUID, err := GetPoolIDWithIOs()
+		log.FailOnError(err, "Failed to get pool using UUID")
+		log.InfoD("Pool UUID on which IO is running [%s]", poolUUID)
+
+		nodeDetail, err := GetNodeWithGivenPoolID(poolUUID)
+		log.FailOnError(err, "Failed to get Node Details from PoolUUID [%v]", poolUUID)
+
+		// Resize Pool with lower pool size than existing
+		// Do Pool Expand on the Node
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6349]: do we need line line number 6349 ?

File: tests/basic/storage_pool_test.go [hunk 12]
```diff
@@ -6315,3 +6315,68 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+
+			log.InfoD("Current Size of the pool [%s] is [%d]", poolUUID, poolToBeResized.TotalSize/units.GiB)
+
+			// Now trying to Expand Pool with reduced Pool size
+			err = Inst().V.ExpandPoolUsingPxctlCmd(*nodeDetail, poolUUID, api.SdkStoragePool_RESIZE_TYPE_AUTO, expectedSize)
+
+			// Verify error on pool expansion failure
+			var errMatch error
+			errMatch = nil
+			re := regexp.MustCompile(fmt.Sprintf("service pool expand: pool: %s is already at a size..*", poolUUID))
```
Comment (vprabhakar-px) [inline tests/basic/storage_pool_test.go:6367]: update the regex : from "service pool expand: pool: %s is already at a size..*" to "service pool expand.*pool.*%s is already at a size.*"

**Review by vprabhakar-px — CHANGES_REQUESTED** [review]

File: tests/basic/storage_pool_test.go (hunk 10 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go:6356]: this is required as pool to resize we are getting here

File: tests/basic/storage_pool_test.go (hunk 11 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go:6349]: Done

File: tests/basic/storage_pool_test.go (hunk 10 above)
Comment (vprabhakar-px) [reply tests/basic/storage_pool_test.go:6356]: why do you need 2 log.failOnError ? on 6354 and 6356 ? for the same error ?

File: tests/basic/storage_pool_test.go (hunk 12 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go:6367]: this is standard message px so the message is exactly so. I don't think it has to be changed

File: tests/basic/storage_pool_test.go [hunk 13]
```diff
@@ -6315,3 +6315,62 @@ var _ = Describe("{PoolResizeInvalidPoolID}", func() {
...
+			}
+			dash.VerifyFatal(errMatch, nil, "Pool expand to lower size than existing pool size completed?")
+
+		})
+	})
+
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+		log.InfoD("Exit from Maintenance mode if Pool is still in Maintenance")
+		log.FailOnError(ExitNodesFromMaintenanceMode(), "exit from maintenance mode failed?")
```
Comment (lsrinivas-pure) [inline tests/basic/storage_pool_test.go:6373]: there is no need for exit maintenance here

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/basic/storage_pool_test.go (hunk 13 above)
Comment (soma-purestorage) [reply tests/basic/storage_pool_test.go]: Done

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Review by vprabhakar-px — APPROVED** [review]

**Pull request merged**